### PR TITLE
Implement MFA verification on CreateAuthenticateChallenge

### PR DESCRIPTION
**Pull request by codingllama**
Implement the capability to check for MFA requirements into CreateAuthenticateChallenge, as described by #32993.

#20343

### CODE REVIEW

**Comment by codingllama** [comment]
Branched from #32993 (proto definitions).

**Comment by codingllama** [comment]
Protobuf linters get confused when we branch from a branch, like this. It'll sort itself once the base is merged.

File: lib/auth/auth.go [hunk 1]
```diff
@@ -5165,17 +5165,39 @@ func (a *Server) ExportUpgradeWindows(ctx context.Context, req proto.ExportUpgra
...
+	case proto.MFARequired_MFA_REQUIRED_NO:
+		return false
+	default: // _UNSPECIFIED or _YES are both treated as required.
+		return true
+	}
+}
+
+func (a *Server) isMFARequired(ctx context.Context, checker services.AccessChecker, req *proto.IsMFARequiredRequest) (resp *proto.IsMFARequiredResponse, err error) {
+	// Assign Required as a function of MFARequired.
+	defer func() {
```
Comment (strideynet) [inline lib/auth/auth.go:5181]: It feels perhaps a little unintuitive to do this as a `defer` - I could see myself coming to this function which is already fairly long and missing the defer and being confused about why it behaves the way it does.

What do you think ?

**Review by strideynet — APPROVED** [review]
Looks good - just one query on the use of a defer.

File: lib/auth/auth.go (hunk 1 above)
Comment (codingllama) [reply lib/auth/auth.go:5181]: The idea behind the defer was to make sure that Required is always a function of MFARequired. I can move to manual assignments if y'all think that's best.

Comment (ibeckermayer) [reply lib/auth/auth.go:5181]: This suggests to me unnecessary redundancy. What if we deprecate the `Required` field and do something like this (in `api/client/proto/types.go`)?

```go
func (m *IsMFARequiredResponse) IsRequired() bool {
	switch m.MFARequired {
	case MFARequired_MFA_REQUIRED_NO:
		return false
	default: // _UNSPECIFIED or _YES are both treated as required.
		return true
	}
}
```

Comment (codingllama) [reply lib/auth/auth.go:5181]: There is some duplication here, although it didn't seem to me like a clear deprecation scenario - some places do want a boolean-like decision, while others care about _UNSPECIFIED. I try to pick my battles when tangling with these parts, so it that's OK I'd rather not pursue this particular tangent right now.

**Review by ibeckermayer — APPROVED** [review]

File: lib/auth/auth.go (hunk 1 above)
Comment (codingllama) [reply lib/auth/auth.go:5181]: Discussed out-of-band, I'll take a TODO with the intent to follow up after my next few PRs. Meanwhile I'll keep the defer. Thanks both and let me know you'd rather see more changes.

**Pull request merged**